src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -629,34 +646,51 @@ async def coursier_fetch_lockfile(lockfile: CoursierResolvedLockfile) -> Resolve
 
 @rule
 async def select_coursier_resolve_for_targets(
-    targets: Targets, jvm: JvmSubsystem
+    coarsened_targets: CoarsenedTargets, jvm: JvmSubsystem
 ) -> CoursierResolveKey:
-    encountered_resolves: list[str] = []
-    encountered_compatible_resolves: list[str] = []
-    for tgt in targets:
-        if tgt.has_field(JvmResolveField):
-            encountered_resolves.extend(jvm.resolves_for_target(tgt))
-        if tgt.has_field(JvmCompatibleResolvesField):
-            encountered_compatible_resolves.extend(jvm.resolves_for_target(tgt))
-
-    # TODO: validate that all resolves are compatible with each other. Note that we don't validate
-    #  that dependencies are compatible, just the specified targets.
-
-    if len(encountered_resolves) > 1:
-        raise AssertionError(
-            "Encountered >1 `resolve` field, which was not expected: "
-            f"{sorted(tgt.address for tgt in targets)}"
+    """Selects and validates (transitively) a single resolve for a set of roots in a compile graph.
+
+    In most cases, a `CoursierResolveKey` should be requested for a single `CoarsenedTarget` root,
+    which avoids coupling un-related roots unnecessarily. But in other cases, a single compatible
+    resolve is required for multiple roots (such as when running a `repl` over unrelated code), and
+    in that case there might be multiple CoarsenedTargets.
+    """
+    root_targets = [t for ct in coarsened_targets for t in ct.members]
+
+    # Find the set of resolves that are compatible with all roots by ANDing them all together.
+    compatible_resolves: set[str] | None = None
+    for tgt in root_targets:
+        current_resolves = set(jvm.resolves_for_target(tgt))
+        if compatible_resolves is None:
+            compatible_resolves = current_resolves
+        else:
+            compatible_resolves &= current_resolves
+
+    # Select a resolve from the compatible set.
+    if not compatible_resolves:
+        raise NoCompatibleResolve(
+            jvm, "The selected targets did not have a resolve in common", root_targets
         )
-    elif len(encountered_resolves) == 1:
-        resolve = encountered_resolves[0]
-    elif encountered_compatible_resolves:
-        resolve = min(encountered_compatible_resolves)
     else:
-        raise AssertionError(
-            f"No `resolve` or `compatible_resolves` specified for these targets: "
-            f"{sorted(tgt.address for tgt in targets)}"
+        # Take the first compatible resolve.
+        resolve = min(compatible_resolves)
```

Comment:
Indentation seems off on this?

Reply:
Hm, it was accurate: it was inside the `else:`. But it doesn't need to be inside the `else`: will just make it unconditional.